🔒make gitlab credentials idempotent 🔒

burigolucas opened the pull request:
Store Gitlab credentials in a secret to address issue #69 

The user can set the desired credentials during installation, otherwise the credentials will be generated. During upgrade, the credentials are retrieved from the secret.

Note that this fix does not attempt to address the issue when the user tries to reset the Gitlab credentials during an upgrade.

eformat commented:
hi @burigolucas .. i have tested this out now. i isolated the Chart so i was only installing gitlab:

```bash
helm upgrade --install gitlab . --namespace tl500-gitlab
```

then tried reinstalling 2 times - the second time i get a failure in postgres pod

```bash
helm upgrade --install gitlab . --namespace tl500-gitlab
Release "gitlab" has been upgraded. Happy Helming!
NAME: gitlab
LAST DEPLOYED: Sun Nov 17 11:46:13 2024
NAMESPACE: tl500-gitlab
STATUS: deployed
REVISION: 3
TEST SUITE: None
```

![image](https://github.com/user-attachments/assets/ee94d06a-723b-4e17-862e-b62daf7d0f48)

```bash
$ oc logs gitlab-ce-postgresql-3-8vfxv
pg_ctl: another server might be running; trying to start server anyway
waiting for server to start....LOG:  redirecting log output to logging collector process
HINT:  Future log output will appear in directory "pg_log".
 done
server started
/var/run/postgresql:5432 - accepting connections
=> sourcing /usr/share/container-scripts/postgresql/start/set_passwords.sh ...
ERROR:  role "4eZBwBAqko" does not exist
```

seems its trying to set a non existent user/role on redeploy ?

burigolucas commented:
Thank you @eformat  for testing this. I have identified and fixed the cause of the problem. In addition, I have identified several other issues and addressed them in the latest commit. Below are the changes:

- GitLab credentials are stored in immutable secret and reused during helm upgrades
- On a event that the user attempts to modify a GitLab credential on helm upgrade, it will fail
- DeploymentConfigs now only refer to values from the ImageStreams as defined in the values file. There is no hard-coded name of ImageStream or ImageStreamTag. This fixes an issue in case the user attempted to use a different app_name instead of gitlab-ce
- Secrets are no longer exposed on plain text when retrieving the DeploymentConfigs objects. Instead, the values are referred from the GitLab credentials secret
- Image used for the Job configure-postgresql refers to the image as defined in the ImageStreams instead of using a hard-coded image reference

Please, note that values.yaml has braking changes.

eformat commented:
thanks a lot for working on this .. testing was all good .. love the `"helm.sh/resource-policy": "keep"` 
cheers Mike
